crates/wasi-threads/src/lib.rs

```diff
@@ -0,0 +1,138 @@
+//! Implement [`wasi-threads`].
+//!
+//! [`wasi-threads`]: https://github.com/WebAssembly/wasi-threads
+
+use anyhow::{anyhow, Result};
+use rand::Rng;
+use std::thread;
+use wasi_common::I32Exit;
+use wasmtime::{Caller, Linker, Module, SharedMemory, Store};
+
+// This name is a function export designated by the wasi-threads specification:
+// https://github.com/WebAssembly/wasi-threads/#detailed-design-discussion
+const WASI_ENTRY_POINT: &str = "wasi_thread_start";
+
+pub struct WasiThreadsCtx<T> {
+    module: Module,
+    linker: Linker<T>,
+}
+
+impl<T: Clone + Send + 'static> WasiThreadsCtx<T> {
+    pub fn new(module: Module, linker: Linker<T>) -> Self {
+        Self { module, linker }
+    }
+
+    pub fn spawn(&self, host: T, thread_start_arg: i32) -> Result<i32> {
+        let module = self.module.clone();
+        let linker = self.linker.clone();
```

Comment:
If possible I think this would be best avoided because a `Linker<T>` is a costly thing to clone.

One option is to wrap it up in an `Arc`, and another would be to thread the `get_cx` callback through to `spawn` so the `host` argument could be projected into a `&WasiThreadsCtx<T>`